### PR TITLE
feat: replace `ProtocolName` with `AsRef<str>`

### LINKED ISSUE #2831: misc/multistream-select: Treat protocol as `String` and not `[u8]`
## Description

Today we treat protocols negotiated via `multistream-select` as a `[u8]`. Instead we should treat it as a `String`.

## Motivation

- I am not aware of a project using non-string protocol identifiers.
- The [specification treats them as strings](https://github.com/libp2p/specs/blob/master/connections/README.md#protocol-negotiation).
- `libp2p-identify` treats protocols as strings. See [spec](https://github.com/libp2p/specs/tree/master/identify#the-identify-message).
- The mismatch is unnecessary.
- See this confusion in the wild https://github.com/libp2p/rust-libp2p/pull/2798#issuecomment-1209026177
- The [Go implementation](https://github.com/multiformats/go-multistream/blob/46ef6cb8dd97068d755228615c74fbd3527fc2d6/client.go#L34) treats protocols as strings. 

## Current Implementation

https://github.com/libp2p/rust-libp2p/blob/3d3666e1a69c82afa9c966d668c32aae36c87599/misc/multistream-select/src/protocol.rs#L69-L77

## Are you planning to do it yourself in a pull request?

No

### PR DESCRIPTION
**Pull request by thomaseizinger**
## Description

Previously, a protocol could be any sequence of bytes as long as it started with `/`. Now, we directly parse a protocol as `String` which enforces it to be valid UTF8.

To notify users of this change, we delete the `ProtocolName` trait. The new requirement is that users need to provide a type that implements `AsRef<str>`.

We also add a `StreamProtocol` newtype in `libp2p-swarm` which provides an easy way for users to ensure their protocol strings are compliant. The newtype enforces that protocol strings start with `/`. `StreamProtocol` also implements `AsRef<str>`, meaning users can directly use it in their upgrades.

`multistream-select` by itself only changes marginally with this patch. The only thing we enforce in the type-system is that protocols must implement `AsRef<str>`.

Resolves: #2831.

## Notes & open questions

This is an alternative to https://github.com/libp2p/rust-libp2p/pull/3745.

I am hoping that with this change, users will already accustom to use the `StreamProtocol` newtype. Regarding https://github.com/libp2p/rust-libp2p/issues/2863, I think it makes the most sense to remove all these traits in one go and directly have the handler return a list of `StreamProtocol`s.

## Dependencies

- https://github.com/libp2p/rust-libp2p/pull/3837
- https://github.com/libp2p/rust-libp2p/pull/3841
- https://github.com/libp2p/rust-libp2p/pull/3842

## Change checklist

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] A changelog entry has been made in the appropriate crates

### CODE REVIEW

**Comment by thomaseizinger** [comment]
@mxinden please review despite broken build. `multistream-select` by itself passes.

**Comment by thomaseizinger** [comment]
I am likely to go for the other solution.

File: core/src/upgrade/apply.rs [hunk 1]
```diff
@@ -261,51 +239,3 @@ where
         }
     }
 }
-
-type NameWrapIter<I> = iter::Map<I, fn(<I as Iterator>::Item) -> NameWrap<<I as Iterator>::Item>>;
-
-/// Wrapper type to expose an `AsRef<[u8]>` impl for all types implementing `ProtocolName`.
-#[derive(Clone)]
-struct NameWrap<N>(N);
```
Comment (mxinden) [inline core/src/upgrade/apply.rs:269]: :tada:

File: core/src/upgrade/denied.rs [hunk 2]
```diff
@@ -29,7 +29,7 @@ use void::Void;
 pub struct DeniedUpgrade;
 
 impl UpgradeInfo for DeniedUpgrade {
-    type Info = &'static [u8];
+    type Info = &'static str;
```
Comment (mxinden) [inline core/src/upgrade/denied.rs:32]: Bummer that `Void` does not implement `AsRef<_>`.

File: swarm/src/stream_protocol.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+use std::fmt;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+
+/// Identifies a protocol for a stream.
+///
+/// libp2p nodes use stream protocols to negotiate what to do with a newly opened stream.
+/// Stream protocols are string-based and must start with a forward slash: `/`.
+#[derive(Debug, Clone, Eq)]
+pub struct StreamProtocol {
```
Comment (mxinden) [inline swarm/src/stream_protocol.rs:11]: To be nit-picky, one could as well use `StreamProtocol` to upgrade a connection, e.g. with a security upgrade like noise. That said, I don't have a better naming suggestion.

Another nit-pick, `StreamProtocolName` would be in my eyes more descriptive, though more verbose. I am fine either way.

**Review by mxinden — DISMISSED** [review]
Change looks good to me. Thanks for going through the whole code-base. (Good that we have a mono-repo.)

I like the change of plan from https://github.com/libp2p/rust-libp2p/pull/3745 to this.

File: swarm/src/stream_protocol.rs (hunk 3 above)
Comment (thomaseizinger) [reply swarm/src/stream_protocol.rs:11]: > To be nit-picky, one could as well use `StreamProtocol` to upgrade a connection, e.g. with a security upgrade like noise. That said, I don't have a better naming suggestion.

I did think about that, however I consider this to be the first step towards separating the use of multistream-select for transport upgrades from stream upgrades.

With increased usage of QUIC, transport upgrades will no longer be needed but stream upgrades will remain. Thus I figured it is good to create an abstraction for stream upgrades that is separate. Note that `StreamProtocol` lives in `libp2p-swarm` whereas all transport upgrades live in `libp2p-core`.

> Another nit-pick, `StreamProtocolName` would be in my eyes more descriptive, though more verbose. I am fine either way.

It is a bit too long for my taste. In discussions, we never talk about a protocol _name_, we talk about stream protocols (vs multiaddr protocols for example).

I think I'll leave it as is.

File: core/src/upgrade/denied.rs (hunk 2 above)
Comment (thomaseizinger) [reply core/src/upgrade/denied.rs:32]: We've had success in upstreaming changes in the past, I'll see if I can get that implemented!

Comment (thomaseizinger) [reply core/src/upgrade/denied.rs:32]: Actually, I mixed it up with `Either` but I'll try anyway!

Comment (thomaseizinger) [reply core/src/upgrade/denied.rs:32]: Unfortunately, it seems to be pretty much unmaintained :(

**Comment by thomaseizinger** [comment]
@mxinden, I pushed some more minor patches including changelog entries. I renamed `from_static` to `new`. I want to encourage users to use this constructor and only fall back to `try_from_owned` when absolutely necessary. I figured if they are named after the same convention, they might be considered "equal" which is not the case.

**Review by mxinden — APPROVED** [review]

**Comment by mxinden** [comment]
@thomaseizinger updating https://github.com/mxinden/kademlia-exporter/ right now. Was it a deliberate choice to not implement `TryFrom<String> for StreamProtocol`? If not, I would create a pull request.

**Comment by thomaseizinger** [comment]
> @thomaseizinger updating https://github.com/mxinden/kademlia-exporter/ right now. Was it a deliberate choice to not implement `TryFrom<String> for StreamProtocol`? If not, I would create a pull request.

The need to abstract over the creation of `StreamProtocol` didn't arise so I never thought of it.

What is your usecase? Why can you not use the existing, much more descriptive constructor?

**Comment by mxinden** [comment]
A user can provide a stream protocol name via a configuration file. I am parsing the stream protocol name as a string and would like to call `protocol_name.try_into.expect("something")`. 

> What is your usecase? Why can you not use the existing, much more descriptive constructor?

Works as well. Just a bit more verbose.

**Comment by thomaseizinger** [comment]
> I am parsing the stream protocol name as a string

Perhaps `FromStr` is what we need?

> > What is your usecase? Why can you not use the existing, much more descriptive constructor?
> 
> Works as well. Just a bit more verbose.

A matter of taste for sure, feel free to send a PR! I tend to avoid calling conversion functions explicitly because they are mostly noise.

**Pull request merged**